trin-core/src/portalnet/overlay_service.rs

```diff
@@ -620,12 +620,14 @@ impl<TContentKey: OverlayContentKey + Send> OverlayService<TContentKey> {
         // which will be received in the main loop.
         tokio::spawn(async move {
             let response = match discovery
-                .send_talk_req(destination, protocol, Message::Request(request).to_bytes())
+                .send_talk_req(destination, protocol, Message::from(request).into())
                 .await
             {
-                Ok(talk_resp) => match Message::from_bytes(&talk_resp) {
-                    Ok(Message::Response(response)) => Ok(response),
-                    Ok(_) => Err(OverlayRequestError::InvalidResponse),
+                Ok(talk_resp) => match Message::try_from(talk_resp) {
+                    Ok(message) => match Response::try_from(message) {
```

Comment:
I guess in this case something like `impl TryFrom<Vec<u8>> for Response`? It seems cleaner to abstract away as much type-matching code as we can behind an `impl` since it'll reduce clutter/improve readability in the more "core" code - but there's probably also a point where it can be taken too far

Reply:
The limitation here is that only `Message` is aware of the SSZ encoding for some arbitrary byte array, because the derived SSZ traits on the enum recognize the union type selector. This is not true for the `Request` and `Response` types.